iree/compiler/Dialect/HAL/Target/LLVM/internal/AndroidLinkerTool.cpp

```diff
@@ -111,30 +147,73 @@ class AndroidLinkerTool : public LinkerTool {
     }
     artifacts.libraryFile.close();
 
-    SmallVector<std::string, 8> flags = {
+    SmallVector<std::string> flags = {
         getSystemToolPath(),
 
         // Avoids including any libc/startup files that initialize the CRT as
         // we don't use any of that. Our shared libraries must be freestanding.
+        //
+        // It matters that this flag isn't prefixed with --for-linker=. Doing so
+        // results in a dlopen error: 'cannot locate symbol "main" referenced by
+        // "iree_dylib_foo.so"'
         "-nostdlib",  // -nodefaultlibs + -nostartfiles
 
+        "-o " + artifacts.libraryFile.path,
+    };
+
+    // Link all input objects. Note that we are not linking whole-archive as
+    // we want to allow dropping of unused codegen outputs.
+    for (auto &objectFile : objectFiles) {
+      flags.push_back(objectFile.path);
+    }
+
+    // Since we are using a clang driver, we need to prefix the flags that are
+    // meant to be only interpreted by the linker.
+    SmallVector<std::string> flagsToPrefixForLinker = {
         // Statically link all dependencies so we don't have any runtime deps.
         // We cannot have any imports in the module we produce.
         "-static",
 
-        "-o " + artifacts.libraryFile.path,
+        // Generate a dynamic library (ELF type: ET_DYN), otherwise dlopen()
+        // won't succeed on it. This is not incompatible with -static. The GNU
+        // man page for ld, `man ld`, says the following:
+        //
+        //   -static
+        //       Do not link against shared libraries. [...] This option can be
+        //       used with -shared. Doing so means that a shared library is
+        //       being created but that all of the library's external references
+        //       must be resolved by pulling in entries from static libraries.
+        //
+        // While that much is said in the GNU ld man page, the reality is that
+        // out of ld.bfd, ld.gold and ld.lld, only ld.lld actually implements
+        // that. Meanwhile, ld.bfd interprets -static -shared as just -static,
+        // and ld.gold rejects -static -shared outright as "incompatible".
+        //
+        // So here we are effectively relying on the linker being ld.lld, which
+        // is the case because we are using Android NDK clang, which execs
+        // Android NDK ld, which is ld.lld, see strace results mentioned in
+        // AndroidLinkerTool class comment.
+        "-shared",
+
+        // As seen in the strace results mentioned in the AndroidLinkerTool
+        // class comment, when Android NDK clang execs ld, it passes -pie to it.
+        // ld considers that to be incompatible with -shared (maybe simply
+        // because -pie stands for position independent EXECUTABLE and -shared
+        // means generate an ET_DYN, not an ET_EXEC?), so we have to pass
+        // this flag now to undo -pie.
+        "-no-pie",
```

Comment:
Note: I didn't feel very comfortable having to add this flag seemingly potentially disabling position-independent code, but it see that the resulting file is still "static-pie".  Anyway, I don't seem to have a choice: "-shared" is incompatible with "-pie" and there is no "-pic" flag. It just seems that "-shared" carries its own implication of position independence.

```
$ file /tmp/mobilebert-baseline-tf2-float.tflite_linked_llvm-46c943.so
/tmp/mobilebert-baseline-tf2-float.tflite_linked_llvm-46c943.so: ELF 64-bit LSB shared object, ARM aarch64, version 1 (SYSV), static-pie linked, with debug_info, not stripped
```